### PR TITLE
Test against Ruby 3.2, fix broken things

### PR DESCRIPTION
**Pull request by magni-**
My build for a PR in `ruby-grape/grape-swagger`, which depends on `mustermann`, [failed on Ruby 3.2](https://github.com/ruby-grape/grape-swagger/runs/7422461584?check_suite_focus=true#step:4:9). Digging into the failure, it seems to be caused by a change concerning `ruby2_keywords`:

> Methods taking a rest parameter (like `*args`) and wishing to delegate keyword arguments through `foo(*args)` must now be marked with `ruby2_keywords` (if not already the case). In other words, all methods wishing to delegate keyword arguments through `*args` must now be marked with `ruby2_keywords`, with no exception. This will make it easier to transition to other ways of delegation once a library can require Ruby 3+. Previously, the `ruby2_keywords` flag was kept if the receiving method took `*args`, but this was a bug and an inconsistency. A good technique to find the potentially-missing `ruby2_keywords` is to run the test suite, for where it fails find the last method which must receive keyword arguments, use `puts nil, caller, nil` there, and check each method/block on the call chain which must delegate keywords is correctly marked as `ruby2_keywords`. [[Bug #18625](https://bugs.ruby-lang.org/issues/18625)] [[Bug #16466](https://bugs.ruby-lang.org/issues/16466)]
>
> ```ruby
> def target(**kw)
> end
>
> # Accidentally worked without ruby2_keywords in Ruby 2.7-3.1, ruby2_keywords
> # needed in 3.2+. Just like (*args, **kwargs) or (...) would be needed on
> # both #foo and #bar when migrating away from ruby2_keywords.
> ruby2_keywords def bar(*args)
>   target(*args)
> end
>
> ruby2_keywords def foo(*args)
>   bar(*args)
> end
>
> foo(k: 1)
> ```

There's probably other methods that now need to be marked? I added `head` to the CI test matrix, so those should be caught there.

### CODE REVIEW

**Comment by dentarg** [comment]
@jkowens are you able to add more maintainers here on GitHub? Would be good to have more people being able to approve CI runs

Anyway, I ran the tests in my fork, see https://github.com/dentarg/mustermann/actions/runs/2702649077, looks like this needs more work

**Comment by dentarg** [comment]
Also, @eregon probably knows how to handle this :)

**Comment by magni-** [comment]
> Anyway, I ran the tests in my fork, see https://github.com/dentarg/mustermann/actions/runs/2702649077, looks like this needs more work

Thanks, looks like `ruby2_keywords` is for module/class _instance_ methods, so `ruby2_keywords def self.parse(...)` won't work.

File: mustermann/lib/mustermann/ast/node.rb [hunk 1]
```diff
@@ -91,6 +93,10 @@ def type
         self.class.type
       end
 
+      class << self
+        ruby2_keywords :parse
+      end
```
Comment (eregon) [inline mustermann/lib/mustermann/ast/node.rb:98]: Could you move it just below the declaration of `def self.parse`? Otherwise it's hard to find (in both directions).

**Comment by eregon** [comment]
Yes, slightly more `ruby2_keywords` are needed for Ruby 3.2+.
That changes makes sense, it might be needed in more places.
An alternative here would be to use `payload = nil, **options` instead of `*args` for `def self.parse`'s arguments, because those arguments are passed to `initialize` which takes `payload = nil, **options`. The advantage is this doesn't need the complex (semantically and a bit of cost) marking of `ruby2_keywords`.

**Comment by magni-** [comment]
The build now passes on my fork on 3.2: https://github.com/magni-/mustermann/runs/7430295996?check_suite_focus=true#step:5:15

It turned out the main cause of failures was the removal of `Kernel#=~`. I went with the ugly easy fix (`respond_to?`) but maybe we shouldn't be passing in a non-`String` as a parameter called `string` ?

While grepping for other places that might be missing a `ruby2_keywords`, I ran into these two:
```diff
diff --git mustermann-contrib/lib/mustermann/flask.rb mustermann-contrib/lib/mustermann/flask.rb
index 4caa3b8..848deab 100644
--- mustermann-contrib/lib/mustermann/flask.rb
+++ mustermann-contrib/lib/mustermann/flask.rb
@@ -179,11 +179,11 @@ def self.register_converter(name, converter = nil, &block)
       converter.node_type = :named_splat
     end
 
-    register_converter(:any) do |converter, *strings|
+    register_converter(:any, &-> (converter, *strings) do
       strings              = strings.map { |s| Regexp.escape(s) unless s == {} }.compact
       converter.qualifier  = ""
       converter.constraint = Regexp.union(*strings)
-    end
+    end.ruby2_keywords)
 
     register_converter(:default, converters['string'])
 
diff --git mustermann/lib/mustermann/ast/transformer.rb mustermann/lib/mustermann/ast/transformer.rb
index f21f8f3..bdfe93f 100644
--- mustermann/lib/mustermann/ast/transformer.rb
+++ mustermann/lib/mustermann/ast/transformer.rb
@@ -1,4 +1,5 @@
 # frozen_string_literal: true
+require 'ruby2_keywords'
 require 'mustermann/ast/translator'
 
 module Mustermann
@@ -139,7 +140,7 @@ def track(element)
 
         # turn look ahead buffer into look ahead node
         # @!visibility private
-        def create_lookahead(elements, *args)
+        ruby2_keywords def create_lookahead(elements, *args)
           return elements unless elements.size > 1
           [Node[:with_look_ahead].new(elements, *args, start: elements.first.start, stop: elements.last.stop)]
         end
```

But the build passes without these changes and I'm really unfamiliar with both this codebase and the `ruby2_keywords` method, so I'm not sure if those changes are actually needed/correct.

**Comment by jkowens** [comment]
> @jkowens are you able to add more maintainers here on GitHub? Would be good to have more people being able to approve CI runs
> 
> Anyway, I ran the tests in my fork, see https://github.com/dentarg/mustermann/actions/runs/2702649077, looks like this needs more work

@dentarg I wish I could, but I don't have access to do that. @zzak could you help us out with that?

**Comment by magni-** [comment]
Also, while the specs pass, the coverage checks blows up because of the removal of the deprecated `File.exists?` method, which is referenced by `coveralls:0.7.2`. `coveralls` [fixed this in `0.7.3`](https://github.com/lemurheavy/coveralls-ruby/commit/ad2875a40a0d7ad5a80b5f858485b30089ab3a2a) to use `File.exist?` instead, but we can't upgrade to that (or to a newer version) because of our dependency on `simplecov` (see https://github.com/lemurheavy/coveralls-ruby/pull/172)

```
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Bundler could not find compatible versions for gem "simplecov":
  In Gemfile:
    support was resolved to 0.0.1, which depends on
      coveralls (~> 0.8.23) was resolved to 0.8.23, which depends on
        simplecov (~> 0.16.1)

    support was resolved to 0.0.1, which depends on
      simplecov (~> 0.17.0)
```

**Comment by dentarg** [comment]
> the coverage checks blows up

Looks like it is the sending to Coveralls that doesn't work but that's already the case for the other jobs too (`Coveralls encountered an exception: RestClient::UnprocessableEntity 422 Unprocessable Entity`), thankfully it doesn't fail the job so we can ignore that for now. :)

File: mustermann/lib/mustermann/ast/expander.rb [hunk 2]
```diff
@@ -124,6 +124,8 @@ def error_for(values)
       # @see Mustermann::AST::Translator#expand
       # @!visibility private
       ruby2_keywords def escape(string, *args)
+        return super unless string.respond_to?(:=~)
```
Comment (eregon) [inline mustermann/lib/mustermann/ast/expander.rb:127]: If it's just nil/false, I'd simplify to `return super unless string` or integrate it in the ternary below.
Or are other types seen?

**Comment by eregon** [comment]
@magni- The first change looks wrong to me, because `*strings` isn't used for delegating keyword arguments.
`Regexp.union` does not accept kwargs AFAIK.
This (existing) line seems dodgy:
```ruby
strings              = strings.map { |s| Regexp.escape(s) unless s == {} }.compact
```
What passes a empty hash to that method? That seems wrong. I think such callers should be fixed to not pass any kwargs or hashes.
But also because of this line the `ruby2_keywords` shouldn't be needed.

The second change also looks strange to me, because AFAIK one cannot use a `ruby2_keywords` Hash + explicit kwargs like in:
```ruby
[Node[:with_look_ahead].new(elements, *args, start: elements.first.start, stop: elements.last.stop)]
```
An example which show it does not work:
```
$ ruby -vwe 'def r(*a,**kw); p [a, kw]; end; ruby2_keywords def m(e, *args); r(e, *args, start: 1); end; m(1, 2, 3, kw: 2)'
ruby 3.2.0dev (2022-06-26T06:36:14Z master c2e37c8ff7) [x86_64-linux]
[[1, 2, 3, {:kw=>2}], {:start=>1}]
```

Do you actually have more warnings or error without these 2 `ruby2_keywords` calls?

File: mustermann/lib/mustermann/ast/node.rb [hunk 3]
```diff
@@ -35,8 +35,8 @@ def self.constant_name(name)
       # Helper for creating a new instance and calling #parse on it.
       # @return [Mustermann::AST::Node]
       # @!visibility private
-      def self.parse(*args, &block)
-        new(*args).tap { |n| n.parse(&block) }
+      def self.parse(payload = nil, **options, &block)
+        new(payload, **options).tap { |n| n.parse(&block) }
```
Comment (eregon) [inline mustermann/lib/mustermann/ast/node.rb:39]: Looks good :+1:

File: mustermann/lib/mustermann/ast/expander.rb (hunk 2 above)
Comment (magni-) [reply mustermann/lib/mustermann/ast/expander.rb:127]: From the build failures, it's `Integer`s: https://github.com/magni-/mustermann/runs/7425121555?check_suite_focus=true#step:5:14

e.g.
```
   9) Mustermann::Expander supports combining different pattern styles
     Failure/Error: expander.expand(bar: 23).should be == "/23"
     NoMethodError:
       undefined method `=~' for 23:Integer
     # ./mustermann/lib/mustermann/ast/expander.rb:128:in `escape'
     # ./mustermann/lib/mustermann/ast/expander.rb:99:in `block in expand'
     # ./mustermann/lib/mustermann/ast/expander.rb:99:in `each'
     # ./mustermann/lib/mustermann/ast/expander.rb:99:in `expand'
     # ./mustermann/lib/mustermann/expander.rb:150:in `expand'
     # ./mustermann/spec/expander_spec.rb:26:in `block (2 levels) in <top (required)>'
```

**Comment by magni-** [comment]
Anything else needed for this PR? 🙇🏼

**Review by eregon — APPROVED** [review]
IMHO good to merge
